### PR TITLE
Balance table treats absent entries as zero, never stores zeros

### PR DESCRIPTION
**Pull request by anorth**
This is motivated by #784. I think this is a much more robust way of cleaning up empty accounts.

Minor conflicts with #775, but please let this land first.

### CODE REVIEW

File: actors/builtin/market/market_test.go [hunk 1]
```diff
@@ -1612,6 +1611,21 @@ func TestCronTickDealExpiry(t *testing.T) {
...
+
+		// move the current epoch so that deal is expired
+		rt.SetEpoch(startEpoch + 1000)
+		actor.cronTick(rt)
+		require.EqualValues(t, deal.ClientCollateral, actor.getEscrowBalance(rt, client))
+
+		// client withdraws collateral -> account should be removed as it now has zero balance
+		actor.withdrawClientBalance(rt, client, deal.ClientCollateral, deal.ClientCollateral)
+		actor.assertAccountZero(rt, client)
+	})
```
Comment (anorth) [inline actors/builtin/market/market_test.go:1628]: From #719

File: actors/builtin/market/market_test.go [hunk 2]
```diff
@@ -1715,8 +1729,19 @@ func TestCronTickDealSlashing(t *testing.T) {
...
+						// client balances should not change
+						cLocked := actor.getLockedBalance(rt, client)
+						cEscrow := actor.getEscrowBalance(rt, client)
+						actor.cronTick(rt)
+						require.EqualValues(t, cEscrow, actor.getEscrowBalance(rt, client))
+						require.EqualValues(t, cLocked, actor.getLockedBalance(rt, client))
+					} else {
+						// running cron tick again dosen't do anything
+						actor.cronTickNoChange(rt, client, provider)
+					}
```
Comment (anorth) [inline actors/builtin/market/market_test.go:1744]: From #719

**Comment by anorth** [comment]
FYI @wadealexc

File: actors/util/adt/balancetable.go [hunk 3]
```diff
@@ -98,17 +74,17 @@ func (t *BalanceTable) SubtractWithMinimum(key addr.Address, req abi.TokenAmount
...
 	if err != nil {
 		return err
 	}
-	if !subst.Equals(req) {
-		return errors.New("couldn't subtract the requested amount")
+	if req.GreaterThan(prev) {
+		return xerrors.New("couldn't subtract the requested amount")
 	}
-	return nil
+	return t.Add(key, req.Neg())
```
Comment (Stebalien) [inline actors/util/adt/balancetable.go:87]: This can end up storing zero. Should we delete in this case?

**Comment by Stebalien** [comment]
I assume you meant to store a zero balance, if previously non-zero. What was the actual problem with deleting zero balances?

File: actors/util/adt/balancetable.go (hunk 3 above)
Comment (anorth) [reply actors/util/adt/balancetable.go:87]: Yes!

**Comment by anorth** [comment]
@Stebalien I did not! Good catch, fixed.

The problem previously was that various call sites required the value to be present, but since zeros were not stored, failed if they logically expected a zero to be stored (e.g. b/c deal price was zero).

**Comment by Stebalien** [comment]
> The problem previously was that various call sites required the value to be present, but since zeros were not stored, failed if they logically expected a zero to be stored (e.g. b/c deal price was zero).

Ah, I see. So this is just a smaller patch that pushes that "zero" to the edges. Makes sense.

**Comment by aarshkshah1992** [comment]
@anorth Can we get this in now ? Will unblock #775 .

**Pull request merged**